docs/design.md

```diff
@@ -0,0 +1,52 @@
+# Cyclone Design
+
+Cyclone is a powerful workflow engine and end-to-end pipeline solution implemented with native Kubernetes resources. 
+Cyclone is architectured with a low-level workflow engine that is application agnostic, offering capabilities like
+workflow DAG scheduling, resource lifecycle management.
+
+## Architecture
+
+![Cyclone arch](images/cyclone-arch.png)
+
+### Components
+
+* Cyclone Web: UI of Cyclone helps users to easily use Cyclone.
+* Cyclone Server: API server of Cyclone provides a set of RESTful APIs to operate Cyclone resources.
+* Workflow Controller: Controller of Cyclone workflow engine, controls execution of stages in workflows.
+
+### Kubernetes CRDs
+
+* Stage: Minimum executable unit for workflow, it does some work on input resources and output results.
+* Resource: Input or output of stages.
+* Workflow: Executable DAG graph composes with stages.
+* WorkflowRun: Running record of workflow.
+* WorkflowTrigger: Auto-trigger policy for workflow.
+
+## Implementation
+
+### Workflow Engine
+
+Workflow controller is the main component of workflow engine. It watches various kinds of Cyclone CRDs, and controls the execution of workflows.
+
+Workflow controller creates a pod for each stage to execute, and there are 3 types of containers in the pod:
+
+* Init Containers: Init containers will do some preparation work of input resources before workload containers start.
+* Workload Containers: Workload containers are functional containers, which do the main work for stages.
+* Sidecar Containers: There are 2 common sidecar containers.
+  * Coordinator: Coordinator sidecar is in charge of log collection, artifact collection and notifying resource resolver to handle output resources.
```

Comment:
Another statement I think we can add here is: Workflow controller regards a stage pod completed when the coordinator sidecar container completed.

Reply:
To keep consistent with other items, only need to describe the function of coordinator sidecar, no need to cover the implementation logic. Besides, the sentence is a little long after added this.

Reply:
I think this is a important function of coordinator, not the implementation. Coordinator determines when a stage pod is regarded as completed.